### PR TITLE
Fixed function and unittests for Armstrong number detection

### LINKED ISSUE #40: Implement Armstrong Number Checker
## Problem Statement:

Write a function that checks if a given integer is an Armstrong number of the first kind.

An Armstrong number of the first kind is a number that is the sum of its own digits each raised to the power of the number of digits. 

For example: 153 is an Armstrong number because 1^3 + 5^3 + 3^3 = 1 + 125 + 27 = 153
                      1634 is an Armstrong number because 1^4 + 6^4 + 3^4 + 4^4 = 1 +1296 + 81+ 256 = 1634 

**Examples:**

* **Input:** 153
  **Output:** True

* **Input:** 370
 **Output:** True

* **Input:** 121
  **Output:** False

* **Input:** 100
  **Output:** False

### PR DESCRIPTION
**Pull request by MuqadsaT**
---
name: Armstrong Number Detection Review
about: Adds the `is_armstrong_number` function and unit tests to the project
---

**Assignees:** @MuqadsaT

**Links:** [https://github.com/MIT-Emerging-Talent/ET6-foundations-group-31/issues/40] 
Fixed by https://github.com/MIT-Emerging-Talent/ET6-foundations-group-31/pull/75

This pull request introduces a solution for detecting Armstrong numbers in Python:

* **`is_armstrong_number.py`:**
    - Implements the `is_armstrong_number` function to determine if a given number is an Armstrong number.
    - Includes type hints, docstrings, and error handling.
    
* **`test_is_armstrong_number.py`:**
    - Contains unit tests for the `is_armstrong_number` function, covering:
        - Standard cases (Armstrong and non-Armstrong numbers)
        - Edge cases (zero, 39-digit input number)
        - Defensive tests (negative input, non-integer input, non_numeric input)

**To-do:**

- [ ] Review and provide any feedback.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/is_armstrong_number.py [hunk 1]
```diff
@@ -25,12 +25,12 @@ def is_armstrong_number(n: int) -> bool:
 
     Raises:
         AssertionError: If n is negative.
```
Comment (Akan186) [inline solutions/is_armstrong_number.py]: I recommend adding single quotes around n in the docstring, as this is a common convention to clearly indicate that the exception applies to the variable n, improving the clarity and readability of the documentation.

File: solutions/is_armstrong_number.py [hunk 2]
```diff
@@ -25,12 +25,12 @@ def is_armstrong_number(n: int) -> bool:
...
-        AssertionError: If n is not an integer.
+        TypeError: If n is not an integer.
```
Comment (Akan186) [inline solutions/is_armstrong_number.py]: same as above. add ' '

File: solutions/is_armstrong_number.py [hunk 3]
```diff
@@ -42,14 +42,17 @@ def is_armstrong_number(n: int) -> bool:
     if not isinstance(n, int):
```
Comment (Akan186) [inline solutions/is_armstrong_number.py:43]: instead of assert, rasising a ValueError would be more appropriate for valid input I think.

**Review by PyaeLinn01 — DISMISSED** [review]

File: solutions/tests/test_is_armstrong_number.py [hunk 4]
```diff
@@ -86,22 +95,25 @@ def test_non_armstrong_number_123(self):
...
+    def test_edge_case_10_digit(self):
+        """Tests if a 10-digit large n is an Armstrong number."""
+        self.assertTrue(is_armstrong_number(4679307774))
+
+    def test_edge_case_39_digit(self):
+        """Tests if a 39-digit large n is an Armstrong number."""
+        self.assertTrue(is_armstrong_number(115132219018763992565095597973971522401))
 
     def test_negative_input(self):
         """Tests if TypeError is raised for negative input."""
```
Comment (Akan186) [inline solutions/tests/test_is_armstrong_number.py]: if you did change line 41 from assert to ValueError, then you must change the TypeError to ValueError

Comment (Akan186) [reply solutions/tests/test_is_armstrong_number.py]: Change the AssertionError below this to ValueError

**Review by Akan186 — CHANGES_REQUESTED** [review]
Great job on the code so far! It would be more appropriate to raise a ValueError instead of AssertionError, since ValueError is typically used for invalid input values. Could you update the function to raise a ValueError for negative numbers?

**Comment by MyatCharm** [comment]
All the left unchecked list has been satisfied I guess. I couldn't make this long list of test file. Wonderful

**Review by MyatCharm — DISMISSED** [review]

File: solutions/is_armstrong_number.py (hunk 3 above)
Comment (MuqadsaT) [reply solutions/is_armstrong_number.py:43]: Assertions verify internal assumptions within the function's logic. Since the focus is on the function's internal correctness, assertions are a suitable choice to help in debugging and writing reliable code.

**Comment by MuqadsaT** [comment]
Assertions verify internal assumptions within the function's logic. Since the focus is on the function's internal correctness, assertions are a suitable choice to help in debugging and writing reliable code.

**Comment by Akan186** [comment]
> Assertions verify internal assumptions within the function's logic. Since the focus is on the function's internal correctness, assertions are a suitable choice to help in debugging and writing reliable code.

Ahh- my fault. I assumed that it was for user input. For internal checks and debugging, Assertions are perfect.

**Review by Akan186 — APPROVED** [review]
I misunderstood. It's all good! (It would work either way, I suppose.)
